### PR TITLE
support proxy protocol on load balancers

### PR DESCRIPTION
**Pull request by DockToFuture**
**How to categorize this PR?**

/area networking
/kind enhancement

**What this PR does / why we need it**:
In order to support IP whitelisting (for example with https://github.com/stackitcloud/gardener-extension-acl) the loadbalancer needs to preserve the client source IP address. As not every loadbalancer is fully transparent proxy protocol needs to be enabled on the loadbalancer to preserve the client IP address. This pull requests configures the envoy in the istio ingress gateway to allow connections with and without proxy protocol.

**Which issue(s) this PR fixes**:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```feature operator
Support for proxy protocol is added to the istio ingress gateway to preserve the client source IP addresses.
```

### CODE REVIEW

**Comment by axel7born** [comment]
Is the value `proxyProtocolEnabled` always `true`?
https://github.com/gardener/gardener/blob/e4062538bb2b68ac45353e902756ce66c1374b62/pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml#L1
I'm wondering if now two different things are configurable via this value.

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml [hunk 1]
```diff
@@ -25,4 +25,65 @@ spec:
         - name: proxy_protocol
           typed_config:
             "@type": "type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol"
-{{ end -}}
\ No newline at end of file
+        - name: inner_proxy_protocol
+          typed_config:
+            "@type": "type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol"
+            allow_requests_without_proxy_protocol: true
```
Comment (ScheererJ) [inline pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml:31]: This is somewhat confusing to me. On the abstract level, port `8443` will always use proxy protocol as `apiserver-proxy` uses proxy protocol. However, this would be the "inner" protocol. I would assume that the load balancer might optionally add another proxy protocol layer, i.e. the "outer" protocol. Hence, I would have expected the `allow_requests_without_proxy_protocol` on the other entry. However, I am not sure if this would even make a difference.
Can you please share your thoughts?

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml [hunk 2]
```diff
@@ -25,4 +25,65 @@ spec:
...
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+  labels:
+{{ .Values.labels | toYaml | indent 4 }}
+  name: proxy-protocol-9443
```
Comment (ScheererJ) [inline pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml]: Do we want to put the port or name in here, i.e. `proxy-protocol-9443` vs. `proxy-protocol-sni`? Depending on the outcome, the original `proxy-protocol` filter should also be renamed.
Personally, I prefer semantic names, but have no hard stance here. What I do want is that the names are consistent.

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml [hunk 3]
```diff
@@ -25,4 +25,65 @@ spec:
...
+      value:
+        listener_filters:
+        - name: proxy_protocol
+          typed_config:
+            '@type': type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
+            allow_requests_without_proxy_protocol: true
+        - name: envoy.filters.listener.tls_inspector
+          typed_config:
+            "@type": type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector
+        per_connection_buffer_limit_bytes: 32768
```
Comment (ScheererJ) [inline pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml]: Why is this necessary? How does this relate to the `istio-ingressgateway` envoy filter in file `envoy-filter.yaml?

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml [hunk 4]
```diff
@@ -25,4 +25,65 @@ spec:
...
+---
+apiVersion: networking.istio.io/v1alpha3
+kind: EnvoyFilter
+metadata:
+  labels:
+{{ .Values.labels | toYaml | indent 4 }}
+  name: proxy-protocol-8132
```
Comment (ScheererJ) [inline pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml]: Same as above.

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml [hunk 5]
```diff
@@ -25,4 +25,65 @@ spec:
...
+      operation: MERGE
+      value:
+        listener_filters:
+        - name: proxy_protocol
+          typed_config:
+            '@type': type.googleapis.com/envoy.extensions.filters.listener.proxy_protocol.v3.ProxyProtocol
+            allow_requests_without_proxy_protocol: true
+        - name: envoy.filters.listener.tls_inspector
+          typed_config:
+            "@type": type.googleapis.com/envoy.extensions.filters.listener.tls_inspector.v3.TlsInspector
```
Comment (ScheererJ) [inline pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml:60]: Why do we need this?

File: pkg/component/networking/istio/istio_test.go [hunk 6]
```diff
@@ -232,8 +232,18 @@ var _ = Describe("istiod", func() {
 			return string(data)
 		}
 
-		istioProxyProtocolEnvoyFilter = func() string {
-			data, _ := os.ReadFile("./test_charts/proxyprotocol_envoyfilter.yaml")
+		istioProxyProtocolEnvoyFilter8443 = func() string {
```
Comment (ScheererJ) [inline pkg/component/networking/istio/istio_test.go]: Again, I would vote for semantic naming, i.e. not including the port names, but rather the name of the port.

**Review by ScheererJ — COMMENTED** [review]
Thanks for this change. If operators can enable/disable proxy protocol without causing any other issues this would certainly be very beneficial.

File: pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml (hunk 5 above)
Comment (DockToFuture) [reply pkg/component/networking/istio/charts/istio/istio-ingress/templates/proxy-protocol-envoyfilter.yaml:60]: Required for tls/ssl.

**Review by ScheererJ — APPROVED** [review]
/lgtm
/approve
